From 8cad869c6adba09d256bf02a1eb7bb08729969bb Mon Sep 17 00:00:00 2001
Message-Id: <8cad869c6adba09d256bf02a1eb7bb08729969bb.1423872525.git.feitong.yi@intel.com>
In-Reply-To: <3062c7e6b919c393846ff0e13558e7f648b352ca.1423872525.git.feitong.yi@intel.com>
References: <3062c7e6b919c393846ff0e13558e7f648b352ca.1423872525.git.feitong.yi@intel.com>
From: Durgadoss R <durgadoss.r@intel.com>
Date: Sun, 25 Jan 2015 15:40:25 +0530
Subject: [PATCH 65/95] MUST_REBASE [VPG]: drivers/video/adf: Fix invalid
 pointer usage

Try to obtain the modelist only when the panel is present.
This avoids crashes due to invalid pointer dereference
in dp_pipe_dump_modes().

MUST_REBASE: Since Google ADF framework is not in upstream and also
we cannot have two display drivers, other one being i915, we need to
work on i915 and ADF convergence path before it can be upstreamed.

Issue: GMINL-5443
Change-Id: Ib5524844d6deb173e26ad696bceed1b91ec7ca82
Signed-off-by: Durgadoss R <durgadoss.r@intel.com>
---
 .../video/adf/intel/core/common/dp/gen_dp_pipe.c   |    7 +++++++
 1 file changed, 7 insertions(+)

diff --git a/drivers/video/adf/intel/core/common/dp/gen_dp_pipe.c b/drivers/video/adf/intel/core/common/dp/gen_dp_pipe.c
index e0ea69b..4d2bf8c 100644
--- a/drivers/video/adf/intel/core/common/dp/gen_dp_pipe.c
+++ b/drivers/video/adf/intel/core/common/dp/gen_dp_pipe.c
@@ -138,6 +138,7 @@ void dp_pipe_dump_modes(struct drm_mode_modeinfo *modelist, u32 n_modes)
 {
 	u32 i = 0;
 	struct drm_mode_modeinfo *modeinfo = modelist;
+
 	for (i = 0; i < n_modes; i++) {
 		pr_info("Mode::%d-%d-%d-%d-%d-%d-%d-%d-%d-%d\n",
 			modeinfo->clock, modeinfo->hdisplay,
@@ -297,6 +298,11 @@ static void dp_pipe_get_modelist(struct intel_pipe *pipe,
 	struct dp_pipe *dp_pipe = to_dp_pipe(pipe);
 	struct dp_panel *panel = &dp_pipe->panel;
 
+	if (!dp_pipe->panel_present) {
+		pr_err("DP panel not present\n");
+		return;
+	}
+
 	if (!panel->no_probed_modes) {
 		pr_err("%s call before probe, returning 0 modes\n", __func__);
 		*n_modes = 0;
@@ -305,6 +311,7 @@ static void dp_pipe_get_modelist(struct intel_pipe *pipe,
 
 	*modelist = panel->modelist;
 	*n_modes = panel->no_probed_modes;
+
 	dp_pipe_dump_modes(*modelist, *n_modes);
 	pr_err("%s done, no_modes=%d\n", __func__, (int)*n_modes);
 }
-- 
1.7.9.5

